### PR TITLE
Implementing FCM sendAll and sendMulticast APIs

### PR DESCRIPTION
**Pull request by hiranya911**
Implementing the new APIs for sending messages in batches:

```
BatchResponse sendAll(List<Message>)
BatchResponse sendAll(List<Message>, boolean)
BatchResponse sendMulticast(MulticastMessage)
BatchResponse sendMulticast(MulticastMessage, boolean)
```

go/fcm-multicast for the full proposal.

Also implementing the following refactorings for clarity and modularity:
* Reducing the size of and responsibilities of `FirebaseMessaging` by introducing 2 new classes - `FirebaseMessagingClient` and `InstanceIdClient`.
* All the rpc logic is now implemented in the new classes, and `FirebaseMessaging` is just the public API for them.
* Moved some of the common API client logic to a new `ApiClientUtils` class, so we can share it across packages in the future.

### CODE REVIEW

File: src/main/java/com/google/firebase/messaging/MulticastMessage.java [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+  /**
+   * Creates a new {@link MulticastMessage.Builder}.
+   *
+   * @return A {@link MulticastMessage.Builder} instance.
+   */
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static class  Builder {
```
Comment (chong-shao) [inline src/main/java/com/google/firebase/messaging/MulticastMessage.java]: Extra space here?

File: src/main/java/com/google/firebase/messaging/FirebaseMessagingClient.java [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+  }
+
+  private FirebaseMessagingException createExceptionFromResponse(HttpResponseException e) {
+    MessagingServiceErrorResponse response = new MessagingServiceErrorResponse();
+    if (e.getContent() != null) {
+      try {
+        JsonParser parser = jsonFactory.createJsonParser(e.getContent());
+        parser.parseAndClose(response);
+      } catch (IOException ignored) {
+        // ignored
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/messaging/FirebaseMessagingClient.java]: If this error is thrown (which would only happen if the error response is not json), then we will pass a default `MessagingServiceErrorResponse` to the `newException()` method. In that case the exception will get created with the  code `UNKNOWN_ERROR` and a default error message containing the full response payload. This behavior is correct for this scenario.

File: src/main/java/com/google/firebase/messaging/InstanceIdClient.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+  }
+
+  private FirebaseMessagingException createExceptionFromResponse(HttpResponseException e) {
+    InstanceIdServiceErrorResponse response = new InstanceIdServiceErrorResponse();
+    if (e.getContent() != null) {
+      try {
+        JsonParser parser = jsonFactory.createJsonParser(e.getContent());
+        parser.parseAndClose(response);
+      } catch (IOException ignored) {
+        // ignored
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/messaging/InstanceIdClient.java]: Same as in `FirebaseMessagingClient`

File: src/main/java/com/google/firebase/messaging/Message.java [hunk 4]
```diff
@@ -79,6 +80,55 @@ private Message(Builder builder) {
...
+  String getTopic() {
+    return topic;
+  }
+
+  @VisibleForTesting
+  String getCondition() {
+    return condition;
+  }
+
+  Map<String, Object> wrapForTransport(boolean dryRun) {
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/main/java/com/google/firebase/messaging/Message.java]: This actually does more than setting `dryRun`. It wraps both the `Message` and the `dryRun` value in a new `Map` and returns the result. This `Map` later gets serialized into JSON when calling the FCM API.

File: src/main/java/com/google/firebase/messaging/MulticastMessage.java (hunk 1 above)
Comment (hiranya911) [reply src/main/java/com/google/firebase/messaging/MulticastMessage.java]: Removed.

**Pull request merged**